### DISCUSSION TITLE
Benefits of using coroutines

### DISCUSSION THREAD

**Discussion by DimitrisCharisis** [ming1/ubdsrv#57] [General]
Hi all!
I have a question: What is the real benefit we gain from using coroutines?
I mean if `handle_io_async` was just preparing the requests and return, and 
then we just (almost) move the code bellow `co_await__suspend_always` into
`tgt_io_done` wouldn't we have the same result?

Thanks

**Comment by ming1** [comment]
> Hi all! I have a question: What is the real benefit we gain from using coroutines? I mean if `handle_io_async` was just preparing the requests and return, and then we just (almost) move the code bellow `co_await__suspend_always` into `tgt_io_done` wouldn't we have the same result?
> 

Hello  Dimitris 

Probably the change you mentioned can work, but the current way is more readable.

`tgt_io_done()` can be simplified to just handle target io completion
by calling io->co.resume() in current implementation.

Meantime all the whole logic for handling IO command can be kept in `handle_io_async()`, the handling could be very complicated, such as, qcow2 needs to run several IOs for handling one IO command. 

So you can get whole picture about how to handle  IO command
from ublk driver, and the logic won't be partitioned to two parts,
one is done in `handle_io_async()`, another part is done in `tgt_io_done()`.

But anyway, it depends on how target code is written, and interfaces
provided by libublksrv should be enough to cover your approach too.

Thanks,
Ming

**Comment by DimitrisCharisis** [comment]
Hi Ming.
Yes this makes sense. I'm using mostly the loop target, so I was wondering if I'm missing something. 
But yes its nice to have to a uniform interface for all targets and this approach seems legit. 

Dimitris
